lib/Infer/ExhaustiveSynthesis.cpp

```diff
@@ -0,0 +1,568 @@
+            VarMap.insert(std::pair<Inst *, llvm::APInt>(Var, NextInput));
+          }
+          if (!HasNextInputValue) break;
+
+          std::map<Inst *, Inst *> InstCache;
+          std::map<Block *, Block *> BlockCache;
+          Inst *Eq =
+            IC.getInst(Inst::Eq, 1,
+                       {getInstCopy(LHS, IC, InstCache, BlockCache, &VarMap, true),
+                        getInstCopy(I, IC, InstCache, BlockCache, &VarMap, true)});
+          Ante = IC.getInst(Inst::And, 1, {Eq, Ante});
+        }
+      }
+      Ante = IC.getInst(Inst::And, 1, {IC.getInst(Inst::Eq, 1, {LHS, I}), Ante});
+      Ante = IC.getInst(Inst::And, 1, {AvoidConsts, Ante});
+      if (DebugLevel > 2) {
+        ReplacementContext RC;
+        RC.printInst(I, llvm::errs(), true);
+
+        llvm::errs() << "\n";
+        RC.printInst(Ante, llvm::errs(), true);
+      }
+      InstMapping Mapping(Ante,
+                          IC.getConst(APInt(1, true)));
+
+      std::vector<Inst *> ModelInsts;
+      std::vector<llvm::APInt> ModelVals;
+      std::string Query = BuildQuery(IC, BPCs, PCs, Mapping, &ModelInsts, /*Negate=*/true);
+
+      bool FirstSmallQueryIsSat;
+      EC = SMTSolver->isSatisfiable(Query, FirstSmallQueryIsSat,
+                                    ModelInsts.size(), &ModelVals, Timeout);
+      if (EC) {
+        return EC;
+      }
+      if (!FirstSmallQueryIsSat) {
+        Unsat++;
+        if (DebugLevel > 2)
+          llvm::errs() << "first query is unsat, all done with this guess\n";
+        continue;
+      }
+      if (DebugLevel > 2)
+        llvm::errs() << "first query is sat\n";
+
+      for (unsigned J = 0; J != ModelInsts.size(); ++J) {
+        if (ModelInsts[J]->Name.find("reserved_") != std::string::npos) {
+          auto Const = IC.getConst(ModelVals[J]);
+          BadConsts[ModelInsts[J]].push_back(Const->Val);
+          auto res = ConstMap.insert(std::pair<Inst *, llvm::APInt>(ModelInsts[J], Const->Val));
+          if (DebugLevel > 2)
+            llvm::errs() << "constant value = " << Const->Val << "\n";
+          if (!res.second)
+            llvm::report_fatal_error("constant already in map");
+        }
+      }
+    }
+
+    BlockPCs BPCsCopy;
+    std::vector<InstMapping> PCsCopy;
+    std::map<Inst *, Inst *> InstCache;
+    std::map<Block *, Block *> BlockCache;
+    auto I2 = getInstCopy(I, IC, InstCache, BlockCache, &ConstMap, false);
+    separateBlockPCs(BPCs, BPCsCopy, InstCache, BlockCache, IC, &ConstMap, false);
+    separatePCs(PCs, PCsCopy, InstCache, BlockCache, IC, &ConstMap, false);
+
+    InstMapping Mapping2(LHS, I2);
+    std::string Query2 = BuildQuery(IC, BPCsCopy, PCsCopy, Mapping2, 0);
+
+    bool SecondSmallQueryIsSat;
+    EC = SMTSolver->isSatisfiable(Query2, SecondSmallQueryIsSat, 0, 0, Timeout);
+    if (EC) {
+      return EC;
+    }
+    if (SecondSmallQueryIsSat) {
+      if (DebugLevel > 2)
+        llvm::errs() << "second query is SAT-- constant doesn't work\n";
```

Comment:
should this be fatal?

Reply:
oops nevermind-- I understand now

Reply:
No this shouldn't be fatal, this only tells the constant C_1 does not fit all input, need to do all the query again with C != C_1